src/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/ParallelRangeManager.cs

```diff
@@ -99,9 +108,24 @@ internal bool FindNewWork(out long nFromInclusiveLocal, out long nToExclusiveLoc
                         Interlocked.CompareExchange(ref _indexRanges[_nCurrentIndexRange]._nSharedCurrentIndexOffset, new Box<long>(0), null);
                     }
 
-                    // this access needs to be on the array slot
-                    long nMyOffset = Interlocked.Add(ref _indexRanges[_nCurrentIndexRange]._nSharedCurrentIndexOffset.Value,
-                                                    _nIncrementValue) - _nIncrementValue;
+                    long nMyOffset;
+                    if (IntPtr.Size == 4 && _use32BitCurrentIndex)
+                    {
+                        // In 32-bit processes, we prefer to use 32-bit interlocked operations, to avoid the possibility of doing
+                        // a 64-bit interlocked when the target value crosses a cache line, as that can be super expensive.
+                        // We use the first 32 bits of the Int64 index in such cases.
+                        unsafe
+                        {
+                            fixed (long* indexPtr = &_indexRanges[_nCurrentIndexRange]._nSharedCurrentIndexOffset.Value)
```

Comment:
Probably, but that would also require adding a reference to the Unsafe library, which I was hesitant to do because this assembly is in netcore's netstandard implementation and Unsafe currently targets netstandard.  I'm not particularly concerned with the fixed here (it's very short), and we can revise it if we like once all of the Unsafe/Span stuff is propagates more throughout corefx.